packages/@ember/-internals/glimmer/lib/renderer.ts

```diff
@@ -281,69 +341,265 @@ interface ViewRegistry {
   [viewId: string]: unknown;
 }
 
-export class Renderer {
-  private _rootTemplate: Template;
-  private _viewRegistry: ViewRegistry;
-  private _roots: RootState[];
-  private _removedRoots: RootState[];
-  private _builder: IBuilder;
-  private _inRenderTransaction = false;
+type Resolver = RuntimeResolver & CompileTimeResolver;
+
+interface RendererData {
+  owner: object;
+  runtime: RuntimeContext;
+  compilation: CompileTimeCompilationContext;
+  builder: (env: Environment, cursor: Cursor) => ElementBuilder;
+  resolver: Resolver;
+  env: {
+    isInteractive: boolean;
+    hasDOM: boolean;
+    document: SimpleDocument;
+  };
+}
 
-  private _owner: InternalOwner;
-  private _context: CompileTimeCompilationContext;
-  private _runtime: RuntimeContext;
+export class RendererState {
```

Comment:
how different is all this from how Ember renders an app? 🤔

Reply:
ah I see: https://github.com/emberjs/ember.js/pull/20781/files#diff-cdf95f0c6c10226a54fbac271478aaa1defd908f0f998fec0f484e52d2544184R603 and https://github.com/emberjs/ember.js/pull/20781/files#diff-cdf95f0c6c10226a54fbac271478aaa1defd908f0f998fec0f484e52d2544184L492 <- original code here, and a lot is added

Reply:
For the record, I think that a lot of this could be factored to share more code. But when I tried to do more than I did here, I kept running into subtle assumptions I was missing in the original code, so I decided to share the obvious stuff and allow more sharing to happen incrementally once people have time to take a careful inventory of the expected behavior of the original code.